### PR TITLE
Add Has WorldQuery

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Implements #3088

### CODE REVIEW

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -553,6 +553,93 @@ impl<'w, 's, T: Component> Fetch<'w, 's> for WriteFetch<T> {
...
+
+    fn update_archetype_component_access(
+        &self,
+        _: &Archetype,
+        _: &mut Access<ArchetypeComponentId>,
+    ) {
+    }
+
+    fn matches_archetype(&self, archetype: &Archetype) -> bool {
+        archetype.contains(self.component_id)
```
Comment (bjorn3) [inline crates/bevy_ecs/src/query/fetch.rs:595]: Isn't the intention that all entities match, but a bool is returned if it has a specific component?

Comment (DJMcNab) [reply crates/bevy_ecs/src/query/fetch.rs:595]: Yeah, the option fetch unconditionally returns true here

File: crates/bevy_ecs/src/query/fetch.rs [hunk 2]
```diff
@@ -553,6 +553,93 @@ impl<'w, 's, T: Component> Fetch<'w, 's> for WriteFetch<T> {
     }
 }
 
+pub struct Has<T: Component>(PhantomData<T>);
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/query/fetch.rs:556]: Why not `Has<Q: WorldQuery>`, having to do `(Has<T>, Has<U>)` and then `&&` the two bools seems annoying

File: crates/bevy_ecs/src/query/fetch.rs [hunk 3]
```diff
@@ -553,6 +553,93 @@ impl<'w, 's, T: Component> Fetch<'w, 's> for WriteFetch<T> {
...
+
+impl<T: Component> WorldQuery for Has<T> {
+    type Fetch = HasFetch<T>;
+    type State = HasState<T>;
+}
+
+pub struct HasFetch<T: Component> {
+    matches: bool,
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply crates/bevy_ecs/src/query/fetch.rs:564]: This field is immutable, so a user can't read it directly. There is also no accessor for it.

[the remark this answers is not in the document]
Comment (BoxyUwU) [reply crates/bevy_ecs/src/query/fetch.rs:564]: The query returns `bool` not `HasFetch`

**Comment by BoxyUwU** [comment]
How is this different than `With<T>`

**Comment by DJMcNab** [comment]
`Has` and `With` are exceptionally similar; the only difference would be the behaviour when used as a filter.

I wonder if a better way to do this would be to create a seperate 'filter_matches_archetype', effectively gaining the current effeciency of `With` for `Has` as a filter (which could be a better name for `With` once they are unified).

Idk, I just don't like that when used as a filter, `Has` will work correctly, but with degraded performance. The current duplication but not quite is not good.

**Comment by cart** [comment]
If we can only have one, I think it needs to be `With` because of the symmetry with `Without`. `Has` and `DoesNotHave` is a no-go.

**Comment by 2ne1ugly** [comment]
Ah well, I reinvented an existing PR again. with 99% matching implementations.
At least I made one simple test. You can take it if needed.
```
    #[test]
    fn has_query() {
        let mut world = World::default();
        let e0 = world.spawn().insert(A(0)).id();
        let e1 = world.spawn().id();

        let ents = world
            .query::<(Entity, Has<A>)>()
            .iter(&world)
            .collect::<Vec<_>>();
        assert_eq!(ents, &[(e1, false), (e0, true)]);
    }
```

**Comment by Nilirad** [comment]
@alice-i-cecile I think this should be closed as well along with #3088.

**Comment by alice-i-cecile** [comment]
Closing for now; we can revive it if stronger use cases come up.

**Pull request closed without merging**